runner/runner.go

```diff
@@ -0,0 +1,75 @@
+package runner
+
+import (
+	"fmt"
+	"time"
+)
+
+type RunId string
+type ProcessState int
+
+const (
+	// An unambiguous 0-value.
+	UNKNOWN ProcessState = iota
+	// Waiting to run.
+	PENDING
+	// Running
+	RUNNING
+
+	// States below are end states
+	// a Process in an end state will not change its state
+
+	// Ran to completion
+	COMPLETED
+	// Could not run to completion
+	FAILED
+)
+
+func (p ProcessState) String() string {
+	switch p {
+	case UNKNOWN:
+		return "UNKNOWN"
+	case PENDING:
+		return "PENDING"
+	case RUNNING:
+		return "RUNNING"
+	case COMPLETED:
+		return "COMPLETED"
+	case FAILED:
+		return "FAILED"
+	default:
+		panic(fmt.Sprintf("Unexpected ProcessState %v", p))
+	}
+}
+
+// A command, execution environment, and timeout.
+type Command struct {
+	// Command line to run
+	Argv []string
+
+	// Key-value pairs for environment variables
+	EnvVars map[string]string
+
+	// Timeout
+	Timeout time.Duration
+}
+
+// Returned by the coordinator when a run request is made.
+type ProcessStatus struct {
+	RunId RunId
+	State ProcessState
+	// References to stdout and stderr, not their text
+	StdoutRef string
+	StderrRef string
+
+	// Only valid if State == COMPLETED
+	ExitCode int
+
+	// Only valid if State == FAILED
+	Error string
+}
+
+type Runner interface {
+	Run(cmd *Command) (*ProcessStatus, error)
```

Comment:
Sure:
1) you get the RunID
2) it could have capacity and have already started running it
3) it could have failed it, because, e.g., you gave it a nonsensical Argv
4) there are other reasons it could fail the request before even returning